substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java

```diff
@@ -226,14 +227,17 @@ protected static ResolvedJavaType getJavaType(HostedField hostedField, boolean w
             if (originalType != null) {
                 return originalType;
             }
-            /* Check for replacement of the original field only. */
-            ResolvedJavaField javaField = hostedField.wrapped.wrapped;
-            if (javaField instanceof SubstitutionField) {
-                return ((SubstitutionField) javaField).getOriginal().getDeclaringClass();
-            }
-            return hostedType.getWrapped().getWrapped();
         }
         ResolvedJavaField javaField = hostedField.wrapped.wrapped;
+        /* If this is a SubstitutionMethod we might need the target not the original */
+        if (javaField instanceof SubstitutionField) {
+            SubstitutionField substitutionField = (SubstitutionField)javaField;
+            if (wantOriginal) {
+                return substitutionField.getOriginal().getDeclaringClass();
+            } else {
+                return substitutionField.getAnnotated().getDeclaringClass();
+            }
+        }
         return javaField.getDeclaringClass();
```

Comment:
I would rename those `getJavaType` methods to `getDeclaringClass`. IMO `getJavaType` is too unspecific (doesn't tell us that we want the type where the method/field is declared in).

Reply:
Done